blobs/p8z77-m_pro/download_BIOS_clean.sh

```diff
@@ -0,0 +1,110 @@
+}
+
+BLOBDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+if [ "$#" -eq 0 ]; then printusage; fi
+
+while getopts ":m:i:" opt; do
+  case $opt in
+    m)
+      if [ -x "$OPTARG" ]; then
+        MECLEAN="$OPTARG"
+      fi
+      ;;
+    i)
+      if [ -x "$OPTARG" ]; then
+        IFDTOOL="$OPTARG"
+      fi
+      ;;
+  esac
+
+done
+
+if [ -z "$MECLEAN" ]; then
+  MECLEAN=`command -v $BLOBDIR/../../build/x86/coreboot-*/util/me_cleaner/me_cleaner.py 2>&1|head -n1`
+  if [ -z "$MECLEAN" ]; then
+    echo "me_cleaner.py required but not found or specified with -m. Aborting."
+    exit 1;
+  fi
+fi
+
+if [ -z "$IFDTOOL" ]; then
+  IFDTOOL=`command -v $BLOBDIR/../../build/x86/coreboot-*/util/ifdtool/ifdtool 2>&1|head -n1`
+  if [ -z "$IFDTOOL" ]; then
+    echo "ifdtool required but not found or specified with -i. Aborting."
+    exit 1;
+  fi
+fi
+
+CAP_ZIP_SHA256SUM="9ea900eccd4a649237b000f1a34beb73cd92fb203d9639d8b7d22ef2a030d360  P8Z77-V-PRO-ASUS-2104.zip"
+CAP_FILE_SHA256SUM="7cf39a893cd6af774e3623a6b80c3e8f8989934b384eff28aba4726e80faa962  P8Z77-V-PRO-ASUS-2104.CAP"
+FINAL_IFD_SHA256SUM="f356c44d1c2b6b574092081bcc5b25eee72effe12f849c62e2725443d41b3b35 flashregion_0_flashdescriptor.bin"
+FINAL_ME_SHA256SUM="5b5685c2b533b9947eced3ff0279e82ad078725ab13245dacfa3a00bed0022dd $BLOBDIR/me.bin"
+FINAL_GBE_SHA256SUM="fca4deb13633712113e1824bfd5afa32f487ca7129ca012fecf5d7502ec1d5ba  flashregion_3_gbe.bin"
+ZIPURL="https://dlcdnets.asus.com/pub/ASUS/mb/LGA1155/P8Z77-V_PRO/P8Z77-V-PRO-ASUS-2104.zip"
+
+ZIPFILENAME=`echo $ZIPURL | sed 's/.*\///'`
+ROMFILENAME=`echo $ZIPFILENAME | sed 's/\.zip$/\.ROM/'`
+
+extractdir=$(mktemp -d)
+echo "### Creating temp dir $extractdir "
+cd "$extractdir"
+
+/bin/cat <<EOF > layout.txt
+00000000:00000fff fd
+000ad000:007fffff bios
+00003000:000acfff me
+00001000:00002fff gb
+EOF
+
+
+echo "### Downloading $ZIPURL"
+wget $ZIPURL || { echo "ERROR: wget failed $ZIPURL" && exit 1; }
+echo "### Verifying expected hash of $ZIPFILENAME"
+echo "$CAP_ZIP_SHA256SUM" | sha256sum --check || { echo "Failed sha256sum verification on downloaded binary..." && exit 1; }
+
+echo "### Extracting Archives"
+unzip $ZIPFILENAME || { echo "Failed unzipping $ZIPFILENAME - Tool installed on host?" && exit 1;}
+
+echo "### Verifying expected hash of $ROMFILENAME"
+echo "$CAP_FILE_SHA256SUM" | sha256sum --check || { echo "Failed sha256sum verification on extracted binary..." && exit 1; }
+
+echo "### extracing BIOS from Capsule"
+dd bs=1024 skip=2 if=P8Z77-V-PRO-ASUS-2104.CAP of=P8Z77-V-PRO-ASUS-2104.ROM || { echo "Failed to de-cap the ROM..." && exit 1; }
+
+echo "### Stock variant AltME & ME Cleaner"
+$IFDTOOL -M 1 $ROMFILENAME
+echo "### extract stock ME"
+$IFDTOOL -x $ROMFILENAME.new
+
+echo "### Applying me_cleaner to neuter and truncate. EFFS,FCRS whitelisted"
+#$MECLEAN -w FCRS,EFFS -t -O "$BLOBDIR/me.bin" flashregion_2_intel_me.bin
+$MECLEAN -t -O "$BLOBDIR/me.bin" flashregion_2_intel_me.bin
```

Comment:
Without relocalisation  of modules some space that could be freed and reused in ifd from me to bios region is lost. Me modules cannot be relocalised?

Reply:
good catch. Ive not really had much need for space crunching on this particular test rig, but yes 

```
8192 Mar 19 15:56 gbe.bin
4096 Mar 19 15:56 ifd.bin
98304 Mar 19 15:56 me.bin
```

Reply:
98304 is a value common to my eyes and expected after relocalization (shrink)

Reply:
> good catch. Ive not really had much need for space crunching on this particular test rig, but yes

That is important. 8mb is really, really limiting and will fail requiring either dropping of platforms at some point or massive collaboration into reducing unneeded features of modules packed into initrd for all boards. That includes massive tweaking of heads kernel and past work to tweak module compilation for space already.

For example, last time I checked, t520 was still using cryptsetup1 and nobody complained. Meaning nobody uses it really and could be dropped.

Goal here being also to lesser maintainer ship burden, which you just accepted to be part of to maintain the life of heads for that platform :)